rules/the-main-principles-of-clean-architecture/rule.md

```diff
@@ -8,24 +8,92 @@ created: 2019-04-14T21:34:38.0000000Z
+
+A common mistake is using a HttpContext directly in the application layer, which is a dependency on the Presentation layer. Instead, the application layer should use an abstraction, or a service using Domain models, which the Presentation layer maps to.
+
+Examples:
+
+  - Application Services
+  - Use Cases/Features
+  - DTOs
+
+### Infrastructure
+
+The Infrastructure layer is where the external systems are interacted with. For example, you might setup a library to wrap a third party Web API or a database. This layer is dependent on the Application layer, but not on anything outside of itself. This layer typically defines the implementation of the interfaces defined in the Application layer.
+
+While this layer might not implement much business logic, it is still important to keep it clean and testable. For general unit testing this layer is the one that is mocked out the most - therefore interfaces should make sense and be easy to mock.
+
+Examples:
+
+  - Persistence
+  - Web API
+  - Email/SMS
+  - Logging
+  - Authentication Provider
+
+### Presentation
+
+The Presentation layer is where the user interface is implemented. This layer is dependent on the Application layer & the Infrastructure layer.
+
+The Presentation layer's sole responsibility is to interface with the UI (e.g. ASP.NET Core Web API, or a GUI) - and map directly to and from the application layer. This layer should not contain any business logic, and should not be dependent on any external items.
+
+The most common use case is a Web API - and it's implementation should define the API routes, its input & output models, using HTTP or other web protocols. The API should then call the application layer, and map the results to the output ViewModel.
+
+## Principles
+
+To achieve the this layering and separation of concerns, we need to follow some principles.
+
+To allow for the separation of concerns, we need to follow the Dependency Inversion Principle (DIP). This principle states that high-level modules should not depend on low-level modules. Both should depend on abstractions. Abstractions should not depend on details. Details should depend on abstractions.
+
+What this means is that the Core layer should not depend on anything outside of itself - and we use interfaces in the Application or Infrastructure layer to achieve this.
+
+Bad Example: Application Depends on Infrastructure, for example relying on a concrete implementation, instead of allowing an interface.
+Good Example: Repository Pattern, where the Application layer depends on an interface, and the Infrastructure layer implements the interface.
+
+Note: A common point of discussion is whether EF Core's DbContext should be passed around, or if it should be wrapped in an interface. You can think of the DbContext as an abstract base class, and the specific connection as the detail. Therefore it is valid to use the DbContext directly, as the class technically does not rely on the specific connection.
```

Comment:
This is a big discussion... I actually agree with this, I think CA should implement the repository pattern and not include the DBContext in the Application layer (as you allude to in your next sentence). See later in the rule:
> Independent of Frameworks - Core should not be dependent on external frameworks such as Entity Framework

But
> whether EF Core's DbContext should be passed around

Could be better worded as:

_whether the Application layer should take a dependency on EF Core, which is an infrastructure concern._

Also, the justification you give is not correct. Having the DbContext in Application breaks the rules of CA; it's not that the connection is the specific implementation and the DbContext is an abstraction, it's that understanding the rules well let's you decide with due consideration when to break them. Therefore, including DbContext in Application has been considered and determined to be an acceptable tradeoff.

Again, as I said, I'm not a fan of it. My personal preference is to remove this note altogether. But I certainly wouldn't object to it being in the rule (I would just prefer it not to be).

Reply:
I agree with you. Similar to CQRS it is outside the scope of the CA rule.